In [1]:
import numpy as np
import pandas as pd

In [2]:
# Importing the takehome_users.csv into pandas dataframe
data1 = pd.read_csv('takehome_users.csv')
df1 = pd.DataFrame(data1)
df1.head()

object_id        creation_time               name  \
0          1  2014-04-22 03:53:30     Clausen August   
1          2  2013-11-15 03:45:04      Poole Matthew   
2          3  2013-03-19 23:14:52  Bottrill Mitchell   
3          4  2013-05-21 08:09:28    Clausen Nicklas   
4          5  2013-01-17 10:14:20          Raw Grace   

                        email creation_source  last_session_creation_time  \
0    AugustCClausen@yahoo.com    GUEST_INVITE                1.398139e+09   
1      MatthewPoole@gustr.com      ORG_INVITE                1.396238e+09   
2  MitchellBottrill@gustr.com      ORG_INVITE                1.363735e+09   
3   NicklasSClausen@yahoo.com    GUEST_INVITE                1.369210e+09   
4          GraceRaw@yahoo.com    GUEST_INVITE                1.358850e+09   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  
0             10803.0  
1               316.0  
2              1525.0  
3              5151.0  
4              5240.0

In [3]:
df1.rename(columns={'object_id':'user_id'}, inplace=True)

In [4]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   user_id                     12000 non-null  int64  
 1   creation_time               12000 non-null  object 
 2   name                        12000 non-null  object 
 3   email                       12000 non-null  object 
 4   creation_source             12000 non-null  object 
 5   last_session_creation_time  8823 non-null   float64
 6   opted_in_to_mailing_list    12000 non-null  int64  
 7   enabled_for_marketing_drip  12000 non-null  int64  
 8   org_id                      12000 non-null  int64  
 9   invited_by_user_id          6417 non-null   float64
dtypes: float64(2), int64(4), object(4)
memory usage: 937.6+ KB


In [5]:
# Converting the datatype of last_session_creation_time from float to datetime
df1['last_session_creation_time'] = pd.to_datetime(df1['last_session_creation_time'])
df1.dtypes

user_id                                int64
creation_time                         object
name                                  object
email                                 object
creation_source                       object
last_session_creation_time    datetime64[ns]
opted_in_to_mailing_list               int64
enabled_for_marketing_drip             int64
org_id                                 int64
invited_by_user_id                   float64
dtype: object

In [6]:
df1.describe().T

count         mean          std  min      25%  \
user_id                     12000.0  6000.500000  3464.245950  1.0  3000.75   
opted_in_to_mailing_list    12000.0     0.249500     0.432742  0.0     0.00   
enabled_for_marketing_drip  12000.0     0.149333     0.356432  0.0     0.00   
org_id                      12000.0   141.884583   124.056723  0.0    29.00   
invited_by_user_id           6417.0  5962.957145  3383.761968  3.0  3058.00   

                               50%      75%      max  
user_id                     6000.5  9000.25  12000.0  
opted_in_to_mailing_list       0.0     0.00      1.0  
enabled_for_marketing_drip     0.0     0.00      1.0  
org_id                       108.0   238.25    416.0  
invited_by_user_id          5954.0  8817.00  11999.0

In [7]:
df1.head()

user_id        creation_time               name  \
0        1  2014-04-22 03:53:30     Clausen August   
1        2  2013-11-15 03:45:04      Poole Matthew   
2        3  2013-03-19 23:14:52  Bottrill Mitchell   
3        4  2013-05-21 08:09:28    Clausen Nicklas   
4        5  2013-01-17 10:14:20          Raw Grace   

                        email creation_source    last_session_creation_time  \
0    AugustCClausen@yahoo.com    GUEST_INVITE 1970-01-01 00:00:01.398138810   
1      MatthewPoole@gustr.com      ORG_INVITE 1970-01-01 00:00:01.396237504   
2  MitchellBottrill@gustr.com      ORG_INVITE 1970-01-01 00:00:01.363734892   
3   NicklasSClausen@yahoo.com    GUEST_INVITE 1970-01-01 00:00:01.369210168   
4          GraceRaw@yahoo.com    GUEST_INVITE 1970-01-01 00:00:01.358849660   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  
0             10803.0  
1               316.0  
2              1525.0  
3              5151.0  
4              5240.0

In [8]:
# Reading data from takehome_user_engagement.csv
data2 = pd.read_csv('takehome_user_engagement.csv')
data2.head()

time_stamp  user_id  visited
0  2014-04-22 03:53:30        1        1
1  2013-11-15 03:45:04        2        1
2  2013-11-29 03:45:04        2        1
3  2013-12-09 03:45:04        2        1
4  2013-12-25 03:45:04        2        1

In [9]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207917 entries, 0 to 207916
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   time_stamp  207917 non-null  object
 1   user_id     207917 non-null  int64 
 2   visited     207917 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 4.8+ MB


In [10]:
# Converting the datatype of time_stamp from object to datetime
data2.time_stamp = pd.to_datetime(data2.time_stamp)
data2.dtypes

time_stamp    datetime64[ns]
user_id                int64
visited                int64
dtype: object

In [11]:
new_data = data2.set_index('time_stamp').groupby('user_id')['visited'].resample('D').count()

In [12]:
# Importing the formatted data into pandas dataframe
df2 = pd.DataFrame(new_data)
df2.head()

visited
user_id time_stamp         
1       2014-04-22        1
2       2013-11-15        1
        2013-11-16        0
        2013-11-17        0
        2013-11-18        0

In [13]:
# Sum the number of visits for seven-rolling-day period
df2 = df2.rolling(window=7, min_periods=1).sum()
df2.reset_index(inplace = True)
df2.head()

user_id time_stamp  visited
0        1 2014-04-22      1.0
1        2 2013-11-15      2.0
2        2 2013-11-16      2.0
3        2 2013-11-17      2.0
4        2 2013-11-18      2.0

In [14]:
# Compute the maximum number of visits for each user for seven-rolling-day period
df3 = df2.groupby('user_id')[['visited']].max()
df3.head()

visited
user_id         
1            1.0
2            3.0
3            2.0
4            3.0
5            4.0

In [15]:
# Adding a new column for adopted user
df3['adopted_user'] = [1 if x >=3 else 0 for x in df3['visited']]
df3.head()

visited  adopted_user
user_id                       
1            1.0             0
2            3.0             1
3            2.0             0
4            3.0             1
5            4.0             1

In [16]:
# Converting the datatype to int
df3['visited'] = df3['visited'].astype(int)
df3['adopted_user'] = df3['adopted_user'].astype(int)

In [17]:
df3.reset_index(inplace=True) 
df3.head()

user_id  visited  adopted_user
0        1        1             0
1        2        3             1
2        3        2             0
3        4        3             1
4        5        4             1

In [18]:
# Merge the two dataframes
final_df = df3.merge(df1, how='inner', on='user_id', sort=True)

In [19]:
final_df.head()

user_id  visited  adopted_user        creation_time               name  \
0        1        1             0  2014-04-22 03:53:30     Clausen August   
1        2        3             1  2013-11-15 03:45:04      Poole Matthew   
2        3        2             0  2013-03-19 23:14:52  Bottrill Mitchell   
3        4        3             1  2013-05-21 08:09:28    Clausen Nicklas   
4        5        4             1  2013-01-17 10:14:20          Raw Grace   

                        email creation_source    last_session_creation_time  \
0    AugustCClausen@yahoo.com    GUEST_INVITE 1970-01-01 00:00:01.398138810   
1      MatthewPoole@gustr.com      ORG_INVITE 1970-01-01 00:00:01.396237504   
2  MitchellBottrill@gustr.com      ORG_INVITE 1970-01-01 00:00:01.363734892   
3   NicklasSClausen@yahoo.com    GUEST_INVITE 1970-01-01 00:00:01.369210168   
4          GraceRaw@yahoo.com    GUEST_INVITE 1970-01-01 00:00:01.358849660   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  
0             10803.0  
1               316.0  
2              1525.0  
3              5151.0  
4              5240.0

In [20]:
final_df['user_id'].nunique()

8823

In [21]:
final_df1 = final_df[final_df['adopted_user'] == 1]
final_df1.head()

user_id  visited  adopted_user        creation_time             name  \
1        2        3             1  2013-11-15 03:45:04    Poole Matthew   
3        4        3             1  2013-05-21 08:09:28  Clausen Nicklas   
4        5        4             1  2013-01-17 10:14:20        Raw Grace   
5        6        5             1  2013-12-17 03:37:06    Cunha Eduardo   
6        7        6             1  2012-12-16 13:24:32     Sewell Tyler   

                           email creation_source  \
1         MatthewPoole@gustr.com      ORG_INVITE   
3      NicklasSClausen@yahoo.com    GUEST_INVITE   
4             GraceRaw@yahoo.com    GUEST_INVITE   
5  EduardoPereiraCunha@yahoo.com    GUEST_INVITE   
6     TylerSewell@jourrapide.com          SIGNUP   

     last_session_creation_time  opted_in_to_mailing_list  \
1 1970-01-01 00:00:01.396237504                         0   
3 1970-01-01 00:00:01.369210168                         0   
4 1970-01-01 00:00:01.358849660                         0   
5 1970-01-01 00:00:01.387424226                         0   
6 1970-01-01 00:00:01.356009872                         0   

   enabled_for_marketing_drip  org_id  invited_by_user_id  
1                           0       1               316.0  
3                           0       1              5151.0  
4                           0     193              5240.0  
5                           0     197             11241.0  
6                           1      37                 NaN

In [22]:
final_df1['user_id'].nunique()

8015

# Conclusion:

Out of 12,000 users registered, we found that nearly 8015 users are adopted users. The features that will be considered for predicting future user adoption can be 
1. No. of active days(difference between creation date and last active session date)  
2. Opted_in_to_mailing_list 
3. Org_id  
4. Invited_by_user_id   

However, these are the factors that I considered while working with it but this can be further developed into a project when we incorporate Machine Learning into it. Using ML models, we can predict the factors that will help in predicting future user adoption with better accuracy.